Warn validators about stake deactivation

cyberbono3 opened the pull request:
|Fixes] #5773

#### Summary of Changes
1. validators_current_stakes
2. ctrlc::set_handler


Questions:
1. Should i match vote_state.node_pubkey and validator public key [here](https://github.com/cyberbono3/solana/blob/warn-validators/core/src/validator.rs#L183)  ?
2. if we vote_state.node_pubkey  == validator.publickey so i can warn this validator in ctrlc::set_handler right?

I need some guidance to proceed further.

mvines commented on core/src/validator.rs:
```diff
@@ -1,4 +1,5 @@
 //! The `fullnode` module hosts all the fullnode microservices.
+extern crate ctrlc;
```
this extern shouldn't be necessary as we're using rust 2018

cyberbono3 commented:
@mvines could you leave a feedback, please?

mvines commented on core/src/validator.rs:
```diff
@@ -169,6 +171,40 @@ impl Validator {
             bank.slots_per_segment(),
         );
 
+        let working_bank = bank_forks.read().unwrap().working_bank();
```
The overall logic here looks good but we'd need it to be turned inside out, in that the stake check and warning should all run in response to `set_handler()`.   The node may not have stake when it boots initially, but sometime later when Ctrl-C is pressed the node may have stake.

mvines commented on core/src/validator.rs:
```diff
@@ -169,6 +171,40 @@ impl Validator {
+                if stake > 0 {
+                    set_handler(move || {
+                        // warn!("Error writing file! {:?}", res);
+                        warn!("Validator with pubkey {:?} has to deactivate {:?} stake", pubkey, stake);
```
It'd be nice to have a counter here too.  The first Ctrl-C warns, the second Ctrl-C terminates (or perhaps only if you Ctrl-C a second time with a couple seconds)

cyberbono3 commented:
@mvines  I added a counter and put logic inside set_handler. Does it look good?

mvines commented on core/src/validator.rs:
```diff
@@ -168,7 +170,44 @@ impl Validator {
         );
-
+        let bf = bank_forks.clone();
+        let kp = keypair.clone();
```
you could just use local variable `id`, declared higher up in this function.  The validator's public key (aka, `id`) will never change at runtime.

mvines commented on core/src/validator.rs:
```diff
@@ -168,7 +170,44 @@ impl Validator {
+        let kp = keypair.clone();
+        let mut counter = 0;
+        let counter_ref = &mut counter;
+        if counter == 0 {
```
I think we want this counter inside the `set_handler()`.  It'll always be `0` here since Validator::new() is only ever called once.  

The point of `counter` is to signal an exit on the 2nd Ctrl+c right?   I don't see any code that checks the value of `counter` (or `counter_ref`) and actually causes an exit.

cyberbono3 replied:
I thought the point of the counter is to warn a validator about stake deactivation upon first press of Ctrl+c and gracefully terminate upon the second press of CTRL+C. That's why I put our counter outside of set_handler() to enforce execution of set_handler only upon first press of Ctrl+C . 

" It'll always be 0 here since Validator::new() is only ever called once."  That's true. i think maybe we have to move set_handler to another place

mvines replied:
I think the current location of `set_handler()` is fine.  It's just that the closure just needs to check whether it's the first or second time it's been executed as it gets executed on every Ctrl+C right?

mvines commented on core/src/validator.rs:
```diff
@@ -168,7 +170,44 @@ impl Validator {
+                .map(|accounts| {
+                    accounts
+                        .iter()
+                        .filter_map(|(pubkey, (stake, account))| {
```
nit: I think we can make this block smaller, like this:
```rust
                        .filter_map(|(pubkey, (stake, account))| {
                            if let Ok(vote_state) = VoteState::deserialize(&account.data) && vote_state.node_pubkey == validator_pubkey { 
                                  return Some((*pubkey, *stake));
                            }
                            None
                        })
```

mvines replied:
Does this approach not work well?

cyberbono3 replied:
What do you mean here?

mvines replied:
I'm just suggesting the `filter_map()` implementation be restructured to avoid the double `else { None }` at the end, with the sample code I wrote

mvines commented on core/src/validator.rs:
```diff
@@ -168,7 +170,44 @@ impl Validator {
+                        .collect()
+                })
+                .unwrap();
+                if current_stakes.len() > 0 {
```
Hmm, I think we could just move the `warn!()` into the `.filter_map()` and then remove this entire block.

mvines commented on core/src/validator.rs:
```diff
@@ -168,7 +170,44 @@ impl Validator {
+        let mut counter = 0;
+        let counter_ref = &mut counter;
+        if counter == 0 {
+            set_handler(move || { 
```
I just realized we'll want to add a config item in `ValidatorConfig` to enable this feature (disabled by default).  Then enable it explicitly in `validator/src/main.rs`.   We need this because there are some tests (the local_cluster/ tests) that create multiple Validators in the same process, and they all can't set a Ctrl+C handler

cyberbono3 replied:
could you give me more context / any code example how to do it?

mvines replied:
Around this area of the code, https://github.com/solana-labs/solana/blob/62c22c6cb19931d228986346fdf2b98fe5bfa27f/validator/src/lib.rs#L432-L437, add a `validator_config.enable_ctrl_c_handler = true` or so.   

Then in https://github.com/solana-labs/solana/blob/d379786c902d319ee739b1c88da822cec9b780b4/core/src/validator.rs#L57-L71, default `enable_ctrl_c_handler` to `false` and that'll take care of the local_cluster use case I mentioned above

cyberbono3 commented on core/src/validator.rs:
```diff
@@ -171,43 +172,37 @@ impl Validator {
-        if counter == 0 {
-            set_handler(move || { 
+        set_handler(move || { 
+                if *counter_ref == 1 {
```
it says: error[E0594]: cannot assign to `*counter_ref`, as `Fn` closures cannot mutate their captured variables . i  googled it, peopel say to use Arc::new(0) in this case. Any suggestions?

mvines replied:
Ah, yeah using an Arc to hold the value sounds good.  Maybe use an atomic as well (`Arc::new(AtomicUsize::new(0));`), not 100% sure if set_handler() would be re-entered if multiple Ctrl+Cs are pressed quickly

cyberbono3 commented:
@mvines ready for review

mvines commented on validator/src/main.rs:
```diff
@@ -437,7 +437,10 @@ fn main() {
     validator_config.rpc_config.drone_addr = matches.value_of("rpc_drone_addr").map(|address| {
-        solana_netutil::parse_host_port(address).expect("failed to parse drone address")
+
+    validator_config.enable_ctrl_c_handler = true; 
```
oops, I think you want this at line 438

mvines commented on core/src/validator.rs:
```diff
@@ -168,7 +173,38 @@ impl Validator {
+                let mut c = counter.lock().unwrap();
+                if *c == 1 {
+                    warn!("Validator exits..");
+                    // run vvalidator exit 
```
`std::process::exit(1);` should do the trick

mvines commented on core/src/validator.rs:
```diff
@@ -168,7 +173,38 @@ impl Validator {
         );
-
+        let bf = bank_forks.clone();
+        let counter = Arc::new(Mutex::new(0));
```
Using an atomic would be nicer here actually:
```rust
let counter = Arc::new(AtomicUsize::new(0));
```
and then we can do this in the handler:
```rust
if counter.fetch_add(1, Ordering::Relaxed) > 0 {
  warn!("Validator aborting on second ^C");
  std::process::exit(1);
}
```.

cyberbono3 commented:
@mvines, thanks for your guidance

cyberbono3 commented:
I just copied programs/bpf/Cargo.lock  from solana master branch. I hope it works.

mvines commented on core/src/validator.rs:
```diff
@@ -168,7 +173,36 @@ impl Validator {
             bank.slots_per_segment(),
         );
-
+        let bf = bank_forks.clone();
```
Can you wrap this entire new block in:
```rust
if config. enable_ctrl_c_handler {
..
}
```
since we only want to execute it when the handler is enabled.   

Once that's done within the new `if` block we can do this, which will avoid the need to use a short less readable `bf` variable name :
```diff
- let bf = bank_forks.clone();
+ let bank_forks = bank_forks.clone();
```

cyberbono3 replied:
Done

mvines commented on validator/src/main.rs:
```diff
@@ -435,9 +435,11 @@ fn main() {
 
     validator_config.rpc_config.drone_addr = matches.value_of("rpc_drone_addr").map(|address| {
-        solana_netutil::parse_host_port(address).expect("failed to parse drone address")
+    
```
can you please avoid making this change

cyberbono3 replied:
If you take a look at line 442, I added this line "solana_netutil::parse_host_port(address).expect("failed to parse drone address")"

mvines commented:
Please feel free to promote this PR from "Draft" to a normal PR.  I think this is almost ready to land!

mvines commented:
> > Also remove the changes to `programs/bpf/Cargo.lock` please.
> 
> I have reviewed programs/bpf/Cargo.lock.
> There is only one line added to programs/bpf/Cargo.lock to enable CTRL+C catch functionality.
> 
> "ctrlc 3.1.3 (registry+https://github.com/rust-lang/crates.io-index)"

Ah, got it.  Yep that's fine thanks :)

mvines commented on core/src/validator.rs:
```diff
@@ -172,14 +174,14 @@ impl Validator {
                     // run vvalidator exit 
                 }
-                *counter_ref += 1;
+                *c += 1;
```
Sorry I'm not quite sure what you mean here?

cyberbono3 replied:
I think this comment has been made last week. It is no longer relevant

cyberbono3 commented:
I ran cargo test in solana/core and got a following [error](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=7639d8e2ad41e4a81901c16f470ff6ef)

cyberbono3 commented:
@mvines, you wrote :
`.map(|accounts| {
                    accounts
                        .iter()
                        .filter_map(|(pubkey, (stake, account))| {`
> Does this approach not work well?

What do you mean by that?

mvines commented on core/Cargo.toml:
```diff
@@ -76,7 +76,8 @@ tokio-codec = "0.1"
 tokio-io = "0.1"
 untrusted = "0.7.0"
-solana-rayon-threadlimit = { path = "../rayon-threadlimit", version = "0.20.0" }
+solana-rayon-threadlimit = { path = "../rayon-threadlimit", version = "0.19.0-pre0" }
```
oops, this change shouldn't have snuck in

mvines replied:
This change is still unneeded, version should remain at 0.20.0 please

mvines commented on core/src/validator.rs:
```diff
@@ -175,7 +180,37 @@ impl Validator {
         );
-
+        let bank_forks = bank_forks.clone();
+        let counter = Arc::new(AtomicUsize::new(0));
```
These two lines can move after `if config. enable_ctrl_c_handler {` to keep their scope as limited as possible

mvines commented on validator/src/main.rs:
```diff
@@ -1,3 +1,3 @@
 fn main() {
     solana_validator::main()
-}
```
Oh where did the change to validator/src/ go that sets `enable_ctrl_c_handler = true`

cyberbono3 commented:
i hope no more errors left. @mvines, could you review it, please?

mvines commented:
Looks super close, thanks for your patience.    I see two issues:
* [ ] Your changes to `validator/src/main.rs` should go into `validator/src/lib.rs` insteado
* [ ] Can you please run `cargo fmt`, which will format the new code you're adding into the standard style

cyberbono3 commented:
@mvines, I applied changes to validator/src/lib.rs  and ran cargo fmt.

mvines commented on validator/src/lib.rs:
```diff
@@ -72,17 +55,19 @@ fn download_tar_bz2(
-        .send()
-        .and_then(|response| response.error_for_status())
-        .map_err(|err| format!("Unable to get: {:?}", err))?;
+    let client = ureq::agent();
```
It looks like you're working off an older version of the code instead of the HEAD of the master branch.  We switched from ureq to reqwest maybe a week ago now, so all these changes shouldn't be present.

cyberbono3 replied:
it seems i just have to copy validator/src/lib.rs of master branch to my branch right?

mvines replied:
I'd get the latest lib.rs from the master branch, then only add in your change (which if I recall is just a single line)

mvines commented on core/src/validator.rs:
```diff
@@ -175,7 +179,32 @@ impl Validator {
+                    std::process::exit(1);
+                }
+                let bank = bank_forks.read().unwrap().working_bank();
+                let _: Vec<_> = bank
```
The point was that if the node currently has no stake it's ok to exit on the first ^C though.  I don't see that code anymore here.   It looks like we now collect all the stake but never check it for 0, and always exit on the second ^C

cyberbono3 commented:
@mvines, i have updated lib.rs ( copied lib.rs from master branch and added one new line) and added few lines that a validator exits on first ^C if he has no stake. Please let me know what I have to do to proceed further.

mvines commented on core/src/validator.rs:
```diff
@@ -175,7 +179,37 @@ impl Validator {
+                        accounts
+                            .iter()
+                            .filter_map(|(pubkey, (stake, account))| {
+                                if let Ok(vote_state) = VoteState::deserialize(&account.data) {
```
We only want to exit if _this_ validator has stake.  Right now this code will exit if any validator has stake.   you can compare `pubkey` with `keypair.pubkey()` and skip over deserializing the vote account of all validators except _this_ one.

cyberbono3 commented on core/src/validator.rs:
```diff
@@ -194,13 +194,12 @@ impl Validator {
+                                if *pubkey == keypair.pubkey() && *stake > 0 {
+                                    if let Ok(vote_state) = VoteState::deserialize(&account.data) {
+                                        warn!("Validator with stake is exiting on first ^C");
                                         std::process::exit(1);
```
@mvines , do u think i should return Some(*pubkey, *stake) somewhere here?

mvines replied:
I'd just return `Some(*stake)` here.   Then turn `epoch_vote_accounts(..).map(|accounts| ...)` into `epoch_vote_accounts(..).filter_map(|accounts| ...)`.  This way once we arrive at line 206 we have either
1. value `None` if the node is not stake, in which can we can `std::process::exit(1)`
2. `Some(stake)` representing the current stake of the node, in which can we can `warn!` the user

cyberbono3 commented:
in /core ran cargo test , it raised the following [error](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a5a2be0a7f466a8635688fa65242b472)

mvines commented:
Hey looks like there's some compile errors with the latest commits

cyberbono3 commented on core/src/validator.rs:
```diff
@@ -187,16 +192,15 @@ impl Validator {
-                    .filter_map(|account| match account {
-                        pubkey, (stake, accoount) => {
+                    .filter_map(|account| match (account) {
+                        Some(pubkey, (stake, account)) => {
```
@mvines, I guess filter_map unwraps an option automatically, so  it is not necessarily to add Some here right?

mvines replied:
Sorry for the review delay here @cyberbono3!  I just realized that we can use [.find()](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.find) instead of `.filter_map()`.  This way `let stakes: Vec<Option<u64>> ...` turns into just `let  stake: Option<u64> ...`, we can remove the `for stake in stakes.iter()` loop entirely and just `match stake { ... }` directly.

cyberbono3 replied:
@mvines thanks for your feedback. I used find instead of filter_map and then matched stake directly. I tried to run cargo test from solana/core to make sure my code has no compile errors, got this [error](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=fca02dcdbe107dd79901eec811f6364c). It seems, that we use outdated librocksdb-sys v5.18.3, the current version is librocksdb-sys = "6.1.3"

cyberbono3 commented:
@mvines, trying to complete this PR i was running cargo test in core/ to find out if my code is correct. I encountered multiple compile [errors](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a136178bb0e5be4a74501864ed81f499) . i have tried to fix them by updeting the version of some libraries  in Cargo.toml. After I fixed these compile errors, other errors have been popped up (e.g. updated library has no such method etc).  i would appreciate your guidance to resolve this issue.

mvines commented on validator/src/lib.rs:
```diff
@@ -457,8 +457,9 @@ pub fn main() {
 
-    validator_config.rpc_config.enable_validator_exit = matches.is_present("enable_rpc_exit");
-
+    validator_config.rpc_config.enable_fullnode_exit = matches.is_present("enable_rpc_exit");
```
This line looks like a minor rebase error, it shouldn't be modified.

cyberbono3 commented on validator/src/lib.rs:
```diff
@@ -457,7 +457,7 @@ pub fn main() {
     validator_config.voting_disabled = matches.is_present("no_voting");
 
-    validator_config.rpc_config.enable_fullnode_exit = matches.is_present("enable_rpc_exit");
+    validator_config.rpc_config.enable_validator_exit = matches.is_present("enable_rpc_exit");
```
fixed rebase error here

cyberbono3 commented on core/src/validator.rs:
```diff
@@ -189,24 +189,19 @@ impl Validator {
+                        account.iter().find(|(pubkey, (stake, account))| {
+                            *pubkey == &keypair.pubkey()
+                                && *stake > 0
+                                && Ok(vote_state) == VoteState::deserialize(&account.data)
```
@mvines   I got this error here ` Ok(vote_state) == VoteState::deserialize(&account.data) ) 
    |                                ^^^^^^^^^^ not found in this scope`   How can I fix it?

mvines replied:
You'll need to bring it into scope with `use solana_vote_api::vote_state::VoteState;`

cyberbono3 replied:
@mvines   it is already [here](https://github.com/cyberbono3/solana/blob/warn-validators/core/src/validator.rs#L33)

CriesofCarrots replied:
Is it maybe the `vote_state` variable that's underlined in the error message?
I think this line can be instead:
```
&& VoteState::deserialize(&account.data).is_ok()
```

cyberbono3 replied:
@CriesofCarrots  thank you

mvines commented:
It would be good to squash the 27 commits here into 1 single commit eventually

cyberbono3 commented:
@mvines i have squashed multiple commits in one

mvines commented:
Hey looks like the rebase isn't quite right.  Look at the diff in the PR, there's a ton of changes that essentially roll back numerous recent commits.   I can't really review this unfortunately.  Please ping me here once you get the rebase right.  Thanks!

mvines commented:
Obsoleted by #6874
